rig/geometry.py

```diff
@@ -214,3 +218,201 @@ def standard_system_dimensions(num_boards):
+
+
+def spinn5_local_eth_coord(x, y, w, h):
+    """Get the coordinates of a chip's local ethernet connected chip.
+
+    Returns the coordinates of the ethernet connected chip on the same board as
+    the supplied chip.
+
+    .. note::
+        This function assumes the system is constructed from SpiNN-5 boards
+
+    Parameters
+    ----------
+    x : int
+    y : int
+    w : int
+        Width of the system in chips.
+    h : int
+        Height of the system in chips.
+    """
+    dx, dy = SPINN5_ETH_OFFSET[y % 12][x % 12]
+    return ((x + dx) % w), ((y + dy) % h)
+
+
+SPINN5_ETH_OFFSET = np.array([
+    [(vx - x, vy - y) for x, (vx, vy) in enumerate(row)]
+    for y, row in enumerate([
+        # Below is an enumeration of the absolute coordinates of the nearest
+        # ethernet connected chip. Note that the above list comprehension
+        # changes these into offsets to the nearest chip.
+        # X:   0         1         2         3         4         5         6         7         8         9        10        11     # noqa Y:
+        [(+0, +0), (+0, +0), (+0, +0), (+0, +0), (+0, +0), (+4, -4), (+4, -4), (+4, -4), (+4, -4), (+4, -4), (+4, -4), (+4, -4)],  # noqa  0
+        [(+0, +0), (+0, +0), (+0, +0), (+0, +0), (+0, +0), (+0, +0), (+4, -4), (+4, -4), (+4, -4), (+4, -4), (+4, -4), (+4, -4)],  # noqa  1
+        [(+0, +0), (+0, +0), (+0, +0), (+0, +0), (+0, +0), (+0, +0), (+0, +0), (+4, -4), (+4, -4), (+4, -4), (+4, -4), (+4, -4)],  # noqa  2
+        [(+0, +0), (+0, +0), (+0, +0), (+0, +0), (+0, +0), (+0, +0), (+0, +0), (+0, +0), (+4, -4), (+4, -4), (+4, -4), (+4, -4)],  # noqa  3
+        [(-4, +4), (+0, +0), (+0, +0), (+0, +0), (+0, +0), (+0, +0), (+0, +0), (+0, +0), (+8, +4), (+8, +4), (+8, +4), (+8, +4)],  # noqa  4
+        [(-4, +4), (-4, +4), (+0, +0), (+0, +0), (+0, +0), (+0, +0), (+0, +0), (+0, +0), (+8, +4), (+8, +4), (+8, +4), (+8, +4)],  # noqa  5
+        [(-4, +4), (-4, +4), (-4, +4), (+0, +0), (+0, +0), (+0, +0), (+0, +0), (+0, +0), (+8, +4), (+8, +4), (+8, +4), (+8, +4)],  # noqa  6
+        [(-4, +4), (-4, +4), (-4, +4), (-4, +4), (+0, +0), (+0, +0), (+0, +0), (+0, +0), (+8, +4), (+8, +4), (+8, +4), (+8, +4)],  # noqa  7
+        [(-4, +4), (-4, +4), (-4, +4), (-4, +4), (+4, +8), (+4, +8), (+4, +8), (+4, +8), (+4, +8), (+8, +4), (+8, +4), (+8, +4)],  # noqa  8
+        [(-4, +4), (-4, +4), (-4, +4), (-4, +4), (+4, +8), (+4, +8), (+4, +8), (+4, +8), (+4, +8), (+4, +8), (+8, +4), (+8, +4)],  # noqa  9
+        [(-4, +4), (-4, +4), (-4, +4), (-4, +4), (+4, +8), (+4, +8), (+4, +8), (+4, +8), (+4, +8), (+4, +8), (+4, +8), (+8, +4)],  # noqa 10
+        [(-4, +4), (-4, +4), (-4, +4), (-4, +4), (+4, +8), (+4, +8), (+4, +8), (+4, +8), (+4, +8), (+4, +8), (+4, +8), (+4, +8)]   # noqa 11
+    ])
+], dtype=int)
+"""SpiNN-5 ethernet connected chip lookup.
+
+Used by :py:func:`.spinn5_local_eth_coord`. Given an x and y chip position
+modulo 12, return the offset of the board's bottom-left chip from the chip's
+position.
```

Comment:
So that lookup would return (-4, -3) for the key (4, 3) since the lookup gives the *offset* of the ethernet connected chip from the current chip. In this case it would indeed point at (0, 0) as the Ethernet connected chip.

I spent *so* long trying to find a closed-form version of this function and falling back on a LUT was very much a last resort (although the one used by ST and Luis).